adapters/repos/db/lsmkv/segment_group_compaction.go

```diff
@@ -24,75 +24,78 @@ import (
+func (sg *SegmentGroup) bestCompactionCandidatePair() []int {
 	sg.maintenanceLock.RLock()
 	defer sg.maintenanceLock.RUnlock()
 
-	// if true, the parent shard has indicated that it has
-	// entered an immutable state. During this time, the
-	// SegmentGroup should refrain from flushing until its
-	// shard indicates otherwise
-	if sg.isReadyOnly() {
-		return false
-	}
-
-	// if there are at least two segments of the same level a regular compaction
-	// can be performed
-
-	levels := map[uint16]int{}
-
-	for _, segment := range sg.segments {
-		levels[segment.level]++
-		if levels[segment.level] > 1 {
-			return true
-		}
+	// Nothing to compact
+	if len(sg.segments) < 2 {
+		return nil
 	}
 
-	return false
-}
-
-func (sg *SegmentGroup) bestCompactionCandidatePair() []int {
-	sg.maintenanceLock.RLock()
-	defer sg.maintenanceLock.RUnlock()
-
 	// first determine the lowest level with candidates
 	levels := map[uint16]int{}
+	lowestPairLevel := uint16(math.MaxUint16)
+	lowestLevel := uint16(math.MaxUint16)
+	lowestIndex := -1
+	// secondLowestLevel := uint16(math.MaxUint16)
+	secondLowestIndex := -1
+	highestLevel := uint16(0)
+	pairExists := false
+	// var lowestSegment *segment
+	// var secondLowestSegment *segment
+
+	for ind, seg := range sg.segments {
+		levels[seg.level]++
+		val := levels[seg.level]
+		if val > 1 {
+			if seg.level < lowestPairLevel {
+				lowestPairLevel = seg.level
+				pairExists = true
+			}
+		}
 
-	for _, segment := range sg.segments {
-		levels[segment.level]++
-	}
-
-	currLowestLevel := uint16(math.MaxUint16)
-	found := false
-	for level, count := range levels {
-		if count < 2 {
-			continue
+		if seg.level < lowestLevel {
+			// secondLowestLevel = lowestLevel
+			// secondLowestSegment = lowestSegment
+			secondLowestIndex = lowestIndex
+			lowestLevel = seg.level
+			// lowestSegment = seg
+			lowestIndex = ind
 		}
 
-		if level < currLowestLevel {
-			currLowestLevel = level
-			found = true
+		if seg.level > highestLevel {
+			highestLevel = seg.level
 		}
 	}
 
-	if !found {
-		return nil
-	}
+	if pairExists {
+		// now pick any two segments which match the level
+		var res []int
 
-	// now pick any two segments which match the level
-	var res []int
+		for i, segment := range sg.segments {
+			if len(res) >= 2 {
+				break
+			}
 
-	for i, segment := range sg.segments {
-		if len(res) >= 2 {
-			break
+			if segment.level == lowestPairLevel {
+				res = append(res, i)
+			}
 		}
 
-		if segment.level == currLowestLevel {
-			res = append(res, i)
+		return res
+	} else {
+		if sg.compactLeftOverSegments {
+			// Some segments exist, but none are of the same level
+			// Merge the two lowest segments
+
+			return []int{secondLowestIndex, lowestIndex}
+		} else {
+			// No segments of the same level exist, and we are not allowed to merge the lowest segments
+			// This means we cannot compact.  Set COMPACT_LEFTOVER_SEGMENTS to true to compact the remaining segments
```

Comment:
I see that the config check for `COMPACT_LEFTOVER_SEGMENTS` was removed, and in place there is a `WithForceCompation` bucket option. However, this bucket opt is unused, and there is no way to set it from the application entrypoint.

I recommend re-adding the env var check, but in the `usecases/config` package where the other env vars are parsed, rather than in `adapters/handlers/rest/configure_api`.  After that, we can make the `config.Enabled` function private again. Exporting this function from the `usecases` pkg into the `adapters` package will eventually lead to a dependency cycle, if it doesn't already.

Then, we can propagate the config setting from the `appState`, down to the db, and finally to the bucket level where it can get set on the segment group.